File: 0801-minimum-swaps-to-make-sequences-increasing/0801-minimum-swaps-to-make-sequences-increasing.js
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number}
 */
var minSwap = function(nums1, nums2) {
  let [swap, noSwap] = [1, 0];
  
  for (let i = 1; i < nums1.length; i++) {
    let [swap2, noSwap2] = [Infinity, Infinity];
    if (nums1[i - 1] < nums1[i] && nums2[i - 1] < nums2[i]) {
      noSwap2 = noSwap;
      swap2 = swap + 1;
    }
    if (nums1[i - 1] < nums2[i] && nums2[i - 1] < nums1[i]) {
      noSwap2 = Math.min(noSwap2, swap);
      swap2 = Math.min(swap2, noSwap + 1);
    }
    
    [swap, noSwap] = [swap2, noSwap2];
  }
  
  return Math.min(swap, noSwap);
};

File: 0801-minimum-swaps-to-make-sequences-increasing/0801-minimum-swaps-to-make-sequences-increasing.py
class Solution:
    def minSwap(self, nums1: List[int], nums2: List[int]) -> int:
        n = len(nums1)
        swap, not_swap = 1, 0
        for i in range(1, n):
            swap2, not_swap2 = n, n
            if nums1[i - 1] < nums1[i] and nums2[i - 1] < nums2[i]:
                not_swap2 = not_swap
                swap2 = swap + 1
            if nums1[i - 1] < nums2[i] and nums2[i - 1] < nums1[i]:
                not_swap2 = min(not_swap2, swap)
                swap2 = min(swap2, not_swap + 1)
            swap, not_swap = swap2, not_swap2
        return min(swap, not_swap)
                
                